src/compiler/checker.ts

```diff
@@ -8190,9 +8205,13 @@ namespace ts {
                 if (container.parent && container.parent.kind === SyntaxKind.ObjectLiteralExpression) {
                     // Note: this works because object literal methods are deferred,
                     // which means that the type of the containing object literal is already known.
-                    const type = checkExpressionCached(<ObjectLiteralExpression>container.parent);
-                    if (type) {
-                        return type;
+                    const contextualType = getContextualType(container.parent as ObjectLiteralExpression);
+                    const literalType = checkExpressionCached(<ObjectLiteralExpression>container.parent);
+                    if (contextualType && literalType) {
+                        return getIntersectionType([contextualType, literalType]);

```

Comment:
You're right, I can't write a test that actually gets the intersection any more. I'll remove it.